### PR TITLE
FOUNDATIONS: Add Artist

### LINKED ISSUE #10: Add Foundation

### PR DESCRIPTION
**Pull request by NaynaMallick**
closes #10

### CODE REVIEW

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            //then
+            await Assert.ThrowsAsync<ArtistDependencyException>(() =>
+                addArtistTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedArtistDependencyException))),
+                        Times.Once);
+
+            this.apiBrokerMock.Verify(broker =>
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: This one goes up, verify in other places too

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            this.apiBrokerMock.Setup(broker =>
+                broker.PostArtistAsync(It.IsAny<Artist>()))
+                    .ThrowsAsync(httpResponseBadRequestException);
+
+            var invalidArtistException =
+                new InvalidArtistException(
+                    httpResponseBadRequestException,
+                        exceptionData);
+
+            var expeectedArtistDependencyError =
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: expectedArtitstDependencyValidationException

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                    expeectedArtistDependencyError))),
+                        Times.Once);
+
+            this.apiBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Theory]
+        [MemberData(nameof(DependencyValidationExceptions))]
+        public async Task ShouldThrowDepedencyValidationExceptionOnAddIfValidationErrorOccursAndLogItAsync(
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: Dependency Typo

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            Artist someArtist = CreateRandomArtist();
+
+            var invalidArtistException =
+                new InvalidArtistException(dependencyValidaionException);
+
+            this.apiBrokerMock.Setup(broker =>
+                broker.PostArtistAsync(It.IsAny<Artist>()))
+                    .ThrowsAsync(dependencyValidaionException);
+
+            var expectedArtistDependencyException =
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: expectedArtistDependencyValidationException

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                httpRequestException,
+                httpReponseUrlNotFoundException,
+                httpResponseUnauthorizedException
+            };
+        }
+
+        public static TheoryData ApiDependencyExceptions()
+        {
+            var responseMessage = new HttpResponseMessage();
+            String exceptionMessage = GetRandomMesaage();
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs]: string

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                    message: exceptionMessage);
+
+            return new TheoryData<Exception>()
+            {
+                httpResponseConflictException,
+                httpResponseFailedDependencyException
+            };
+        }
+
+        private static Dictionary<string, List<string>> CreateRandomDictionary() =>
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs]: Move the Filler method to last

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        private static Expression<Func<Xeption, bool>> SameExceptionAs(Xeption expectedExeption)
+        {
+            return actualException =>
+                actualException.Message == expectedExeption.Message
+                && actualException.InnerException.Message == expectedExeption.InnerException.Message
+                && (actualException.InnerException as Xeption).DataEquals(expectedExeption.InnerException.Data);
+        }
+
+        private static string GetRandomEmail() =>
+          new EmailAddresses().GetValue().ToString();
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs]: Check the tab here

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            CreateArtistFiller().Create();
+
+        private static Filler<Artist> CreateArtistFiller()
+        {
+            var filler = new Filler<Artist>();
+            Guid id = Guid.NewGuid();
+
+            filler.Setup()
+                .OnType<Guid>().Use(id)
+                .OnType<DateTimeOffset>().Use(DateTimeOffset.UtcNow)
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs]: We should use dateTimeBroker here??

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs (hunk 1 above)
Comment (NaynaMallick) [reply ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: resolved

File: ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 9]
```diff
@@ -118,7 +118,7 @@ public async Task ShouldThrowDependencyValidationExceptionOnAddIfBadRequestAndLo
                     httpResponseBadRequestException,
                         exceptionData);
 
-            var expeectedArtistDependencyError =
+            var expeectedArtistDependencyValidationException =
```
Comment (ShriHumrudha) [inline ArtGallery.Web.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: double ee

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

**Review by ShriHumrudha — APPROVED** [review]

**Pull request merged**